components/operator/ScalarProduct.cpp

```diff
@@ -0,0 +1,73 @@
+#include "ScalarProduct.h"
+
+
+ScalarProduct::ScalarProduct(size_t number_of_spins) {
+    coefficients.resize(number_of_spins, number_of_spins);
+    for (int i = 0; i < number_of_spins; ++i) {
+        for (int j = i + 1; j < number_of_spins; ++j) {
+            // TODO: it does not looks good
+            coefficients(i, j) = -1;
+            coefficients(j, i) = -1;
+        }
+    }
+}
+
+ScalarProduct::ScalarProduct(arma::dmat parameters) : coefficients(std::move(parameters)) {
+}
+
+void ScalarProduct::construct(LexicograficalMatrix& matrix_in_lexicografical_basis, const spaces::LexicographicIndexConverter& converter,
+                                             uint32_t index_of_vector, uint32_t center_a, uint32_t center_b) const {
+    if (!std::isnan(coefficients(center_a, center_b))) {
+        double factor = -2 * coefficients(center_a, center_b);
+        scalar_product_total(matrix_in_lexicografical_basis, converter, index_of_vector, center_a, center_b, factor);
```

Comment:
can a method of `matrix_in_lexicografical_basis ` be used here instead of `scalar_product_total ` ?